### PR TITLE
HD Wallet implementation

### PR DESCRIPTION
**Pull request by CaiMusso**
**This PR closes the implementation of the Digital Wallet for Spring 2021.**  

Content proposed in this Pull Request
----------------------------
This branch adds the implementation of all **functions** needed by the **hierarchical deterministic Wallet**, as well as a new updated **Wallet UI:**
+ The user can create a new wallet or sync it's existing wallet. (already in master)
+ The wallet is initialised with a 12-word mnemonic which is translated to a **master key** (seed)
+ The wallet can now **generate a PoP token** (private and public key) starting from the personal seed and the **path** built using LAO and Roll Call Id.
+ The user can now **view all the PoP tokens associated with its seed** directly on the wallet page. Each of these tokens can be **copied** to clipboard using the copy button or **scanned** by pressing "show QR code".
+ The wallet is also available on the LAO UI.

The **whole process of generating and obtaining a PoP-token once the roll call starts, then saving it to state and being able to retrieve it after having disconnected from the PoP App is now complete and part of this Pull Request**. 

This PR and the Cast Vote PR by @Adalsteinnjons together will add all the functionalities showed in the DEMO of the final presentation. 

### Files that significantly changed for Wallet
+ **HDWallet.ts -** the wallet **object**,  added all **hierarchical deterministic Wallet** functionalities (path derivation).
+ **WalletSyncedSeed -** the main **wallet UI**, added functionality to display all the tokens retrieved from the wallet.

### New files
+ **WalletError.tsx -** a page which informs the user that there was a synchronisation error with the wallet (e.g. mistyped the 12-word mnemonic seed).
+ **CopiableTextBlock.tsx -** a component (text block + button). You can copy the text in text block by pressing the copy icon.
+ **CopyButton.tsx -** a component (button) with the copy icon.
+ **LastPopTokenReducer.ts** - the reducer to store the last generated PoP token.
+ **LastPoPTokenStore.ts** - the store slice for the last generated PoP token.

--------------------------------------------------------------------------------------------------------------------------------

This branch also adds the **QR code reader functionality** on the **Connect** tab, the user can now scan the LAO's QR code instead of manually inserting the LAO ID in order to connect to the LAO.

### Files that significantly changed for QRCode reader
+ **ConnectOpenScan.tsx** - contains the actual QR code reader object
+ **ConnectConfirm.tsx** - manages the information passed by the QR reader and connects to LAO.

### Other changes worth mentioning
+ In **fe1-web/components/QRCode.tsx** (the QR code **displayer**) a new props (**value**) was added. This way the QR code component is generic (namely gets the value it displays as a parameter). The codebase was modified (only in fe1-web/parts/lao/Identity.tsx) to match the new component.

New dependencies
-------------------------------------
+ [**ed25519-hd-key**](https://www.npmjs.com/package/ed25519-hd-key) - used to derive a private and public key for a given seed and a given path.
+ [**react-qr-reader**](https://www.npmjs.com/package/react-qr-reader) - used to read the LAO ID from the QR code.

## TEST and Cross-Testing
The **Wallet.test.ts** covers all the procedure of token derivation from the beginning. It **creates** and **initialises** a wallet, then destroys it and re-**creates it from the Redux state**, verifies if it produces the same master key. 
Furthermore the seed, private key and public key are tested after being derived to be sure they produce the same result as the Java front-end when starting with the same 12-word mnemonic, LAO ID and RollCall ID. **The cross-testing is successful.**

Quick demo
-------------------------------

https://user-images.githubusercontent.com/47753346/122795882-100f7600-d2be-11eb-94af-967ae0eded2b.mov

### CODE REVIEW

File: fe1-web/__tests__/model/objects/Wallet.test.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      const testMap: Map<[Hash, Hash], string[]> = new Map();
+
+      const javaFrontEndPopTokenForLao1RC1: string = '7147759d146897111bcf74f60a1948b1d3a22c9199a6b88c236eb7326adc2efc';
+      const javaFrontEndPopTokenForLao2RC2: string = '2c23cfe90936a65839fb64dfb961690c3d8a5a1262f0156cf059b0c45a2eabff';
+
+      testMap.set([laoId1, rollCallId1], [javaFrontEndPopTokenForLao1RC1, '']);
+      testMap.set([laoId2, rollCallId2], ['', javaFrontEndPopTokenForLao2RC2]);
+
+      let isFirstOfTwoTokens = true;
+      wallet.recoverAllKeys(testMap).then((laoAndRcIdsToTokenMap) => {
```
Comment (pierluca) [inline fe1-web/__tests__/model/objects/Wallet.test.ts:81]: The API confuses me a bit here.

**Review by pierluca — CHANGES_REQUESTED** [review]
A bunch of things in the various APIs of the classes (and some implementations) are a bit confusing.
I've tried to highlight them here.

File: fe1-web/components/CopyButton.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+const styles = StyleSheet.create({
+  icon: {
+    width: 26,
+    height: 26,
+    cursor: 'pointer',
+  } as ImageStyle,
+});
+
+function CopyButton({ action }: IPropTypes) {
```
[the remark this answers is not in the document]
Comment (CaiMusso) [reply fe1-web/components/CopyButton.tsx:21]: Yes very good point, thanks

File: fe1-web/components/CopiableTextBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+});
+
+const CopiableTextBlock = (props: IPropTypes) => {
+  const { id } = props;
+  const { text } = props;
+  const { visibility } = props;
+
+  return (
+    <View style={styles.view}>
+      <CopyButton action={() => navigator.clipboard.writeText(text)} />
```
[the remark this answers is not in the document]
Comment (CaiMusso) [reply fe1-web/components/CopiableTextBlock.tsx:30]: Sure right !

**Pull request merged**